SILGen: Fix order of operations when a mutating existential method returns Self.

jckarter opened the pull request:
The Self type doesn't get opened until immediately before the function call in a mutating method invocation on an existential, so we have to defer formation of the result buffer to that point. rdar://problem/43507711

jckarter commented:
@swift-ci Please test

slavapestov commented on lib/SILGen/ResultPlan.cpp:
```diff
@@ -47,10 +48,137 @@ class InPlaceInitializationResultPlan : public ResultPlan {
+    SmallVector<ArchetypeType *, 4> opened;
+    resultTy->getOpenedExistentials(opened);
+    
+    SmallVector<Type, 4> types;
```
I would suggest factoring out the code for building this signature and layout into a new method

slavapestov commented on lib/SILGen/ResultPlan.cpp:
```diff
@@ -47,10 +48,137 @@ class InPlaceInitializationResultPlan : public ResultPlan {
+    // We allocate the buffer as a box because the scope nesting won't clean
+    // this up with good stack discipline relative to any stack allocations that
+    // occur during argument emission. Escape analysis during mandatory passes
+    // ought to clean this up.
```
Would be good to add a test for that. However another approach would be to allocate a stack buffer of existential type and then _open_ the existential late, using the same opened type as the result of the call. Would that make sense? Then at least you avoid the heap allocation if the value fits inline.

rjmccall replied:
This is a case where we're always going to erase the existential after the call, right?  Can we shift that erasure into the result plan so that we emit into the address produced by `init_existential_addr`?  That would also let us actually do initialization into the context if possible.

slavapestov replied:
Yeah, that’s basically what I’m suggesting.

slavapestov commented on include/swift/SIL/SILType.h:
```diff
@@ -484,6 +484,9 @@ class SILType {
   SILType getReferentType(SILModule &M) const;
+  
+  /// Returns a SILType with any archetypes mapped out of context.
+  SILType mapTypeOutOfContext() const;
```
Do you mind splitting this off into a separate commit and doing a quick pass over mapTypeOutOfContext() calls throughout SIL to see if it can be used anywhere else?

slavapestov reviewed (commented):
I know this is WIP, just a few minor comments. @rjmccall will hopefully do a more in-depth review because the ResultPlan stuff was his design.

jckarter commented:
@slavapestov @rjmccall Updated to also form the concrete value buffer for existentials late as well. That doesn't quite cover all the necessary cases, since if the existential is loadable (as when we're calling a method on an opaque protocol from a class-constrained refining existential) we won't pre-allocate the return buffer, and if there's a `Self?` return, we have to do an optional-to-optional conversion from Self to the existential. The special `ResultPlan` is still necessary, but the common existential case should at least not regress in efficiency.

jckarter commented:
@swift-ci Please test

slavapestov commented:
@jckarter An optional-of-opened-existential-to-optional-of-existential conversion could also theoretically be done in place -- emit the optional existential, project the optional payload and existential payload, and pass that as the out parameter. However it's understandable if you don't want to keep tweaking this.

jckarter commented:
I may be misunderstanding you, but `Optional<T>` can be larger than `T`, and in that case writing an `Optional<T>` into an existential buffer would clobber its metadata pointer.

I'm hoping for now to plug enough holes in the regressions here to give us time to figure out how to do the right thing here, opening the type early while still being able to guarantee invariants when the existential is opened for mutation.

jckarter committed:
Introduce SILType::mapTypeOutOfContext.

This does the same thing as taking the AST type and running it through mapTypeOutOfContext, but
saves call sites from having to do the unwrap-rewrap dance.

jckarter committed:
SILGen: Fix order of operations when invoking a mutating method on an existential that returns Self.

Delay allocating the result buffer for an opened Self return until right before it's needed. When a mutating method is invoked on an existential, the Self type won't be opened until late, when the formal access to the mutable value begins. Fixes rdar://problem/43507711.

jckarter commented:
@slavapestov This look OK to commit now?

How would you feel about bringing this to the 4.2 branch? I'm a bit concerned about introducing new regressions with the changed order of operations in existential codegen; it ought to be benign and doesn't appear to impact any tests, but it's still a change.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please benchmark

jckarter commented:
@swift-ci Please benchmark

jckarter commented:
@swift-ci Please benchmark

jckarter commented:
@swift-ci Please benchmark

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please benchmark

slavapestov commented on test/SILGen/existential_erasure_mutating_covariant_self.swift:
```diff
@@ -0,0 +1,73 @@
+
+  // FIXME: the dynamic self capture here has to happen after existential
+  // opening as well.
+  //_ = y.covariantClosureArgAndReturn({ _ in 0 })
```
Is there a bug for this?

jckarter replied:
Yep, I filed rdar://problem/43778354.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test Linux
